pkg/hackernews/client.go

```diff
@@ -0,0 +1,84 @@
+package hackernews
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/emmaLP/gs-software-onboarding/pkg/hackernews/model"
+)
+
+type Client interface {
+	GetTopStories() ([]int, error)
+	GetItem(id int) (*model.Item, error)
+}
+
+type client struct {
+	httpClient *http.Client
+	baseUrl    string
+}
+
+const (
+	topStoriesPath = "%s/topstories.json"
+	itemPath       = "%s/item/%d.json"
+)
+
+func New(baseUrl string, c *http.Client) (*client, error) {
+	if c == nil {
+		c = http.DefaultClient
+	}
+
+	if len(baseUrl) == 0 {
+		return nil, errors.New("baseURL cannot be nil or empty")
+	}
+
+	return &client{httpClient: c, baseUrl: baseUrl}, nil
+}
+
+func (c *client) GetTopStories() ([]int, error) {
+	path := fmt.Sprintf(topStoriesPath, c.baseUrl)
+	var ids []int
+	if err := c.performRequest(path, http.MethodGet, &ids); err != nil {
+		return nil, err
+	}
+
+	return ids, nil
+}
+func (c *client) GetItem(id int) (*model.Item, error) {
+	path := fmt.Sprintf(itemPath, c.baseUrl, id)
+	var item *model.Item
+
+	if err := c.performRequest(path, http.MethodGet, &item); err != nil {
+		return nil, err
+	}
+
+	return item, nil
+}
+
+func (c *client) performRequest(path, method string, result interface{}) error {
+	request, err := http.NewRequest(method, path, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to creae http request object: %w", err)
+	}
+
+	response, err := c.httpClient.Do(request)
+	if err != nil {
+		return fmt.Errorf("Unable to make http call: %w", err)
+	}
+	defer func(Body io.ReadCloser) {
```

Comment:
This could be `defer resp.Body.Close()` rather than creating the anonymous function.

Reply:
So if I do that I get a warning about not handling the error that the `Close()` method returns

Reply:
I've updated the solution to handle the error without trying to use the defer function. 
This article gave some interesting points around defer and handling errors: https://www.joeshaw.org/dont-defer-close-on-writable-files/